include/bl_storage.h

```diff
@@ -7,15 +7,75 @@
 #ifndef BL_STORAGE_H_
 #define BL_STORAGE_H_
 
-#include <zephyr/types.h>
 #include <string.h>
+#include <zephyr/types.h>
+#include <drivers/nrfx_common.h>
+#include <nrfx_nvmc.h>
+#include <errno.h>
+#include <pm_config.h>
+/* In case this gets included in an image that is build without a bootloader
+ * the provision area is not defined, so we have to point to the OTP manually
+ * NCSDK-18032 is to prevent this workaround in the future
+ */
+#ifndef PM_PROVISION_ADDRESS
+#define PM_PROVISION_ADDRESS (&NRF_UICR->OTP)
+#endif
+
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-
 #define EHASHFF 113 /* A hash contains too many 0xFs. */
+#define EREADLCS 114 /* LCS field of OTP is in an invalid state */
+#define EINVALIDLCS 115 /* Invalid LCS*/
+
+/* We truncate the 32 byte sha256 down to 16 bytes before storing it */
+#define SB_PUBLIC_KEY_HASH_LEN 16
+
+/** Storage for the PRoT Security Lifecycle state, that consists of 4 states:
+ *  - Device assembly and test
+ *  - PRoT Provisioning
+ *  - Secured
+ *  - Decommissioned
+ *  These states are transitioned top down during the life time of a device.
+ *  The Device assembly and test state is implicitly defined by checking if
+ *  the provisioning state wasn't entered yet.
+ *  This works as ASSEMBLY implies the OTP to be erased.
+ */
+struct life_cycle_state_data {
+	uint16_t provisioning;
+	uint16_t secure;
+	/* Pad to end the alignment at a 4-byte boundary as the UICR->OTP
+	 * only supports 4-byte reads. We place the reserved padding in
+	 * the middle of the struct in case we ever need to support
+	 * another state.
+	 */
+	uint16_t reserved_for_padding;
+	uint16_t decommissioned;
+};
+
+/** The first data structure in the bootloader storage. It has unknown length
+ *  since 'key_data' is repeated. This data structure is immediately followed by
+ *  struct counter_collection.
+ */
+struct bl_storage_data {
```

Comment:
If we are either way changing the format, I usually prefer to have a "format version" number at the front, that can be changed later if the format changes again. It's an easy way to check that the structure is what you expect. Looking at the PSA Token spec, this might be covered by that. We could place the implementation_id first, and use one of its bytes as this format version. Just an idea.

Reply:
I think that is the purpose of the implementation ID.

We used to not have it at the front because it was conditional.

I suppose we could move it to the front to make checking the version without type/version information easier.

Reply:
In hindsight, this would have been great to have.

But I consider this to be a feature request and not an issue (old bl_storage didn't have a version identifier either), so not fixing this.

Will keep this in mind for the next time this is re-written.

Reply:
I highly recommend doing this now. The reason is, like we already discussed, that until now, bl_storage has been used only by the immutable bootloader, whereas the plan is now to use it in updateable components, which means there will be a much bigger need for continuity.

Reply:
So currently the only users of the library are the provisioning image (only flashed once before the immutable bootloader) and TF-M.  We are planning to rewrite the logic of the provisioning to remove the dependency of the provisioning script. Since TF-M is now experimental I think that it is an reasonable assumption to assume that can be incompatible with later versions. I fully agree that a some sort of identifier for the format is useful I am just worried that the work that we will do now we will have to undo soon when we write the provisioning logic.

Reply:
I realize you don't want feature creep, but to be clear, what I'm proposing is like 5-10 lines of code spread around. A few in provision.py to place a '1' at the beginning, one line in the bl_storage struct definition, and a few checks for `BL_STORAGE->format_version == 1` in strategic locations. If you foresee things changing in the future, even more reason. Small investment for a lack of headache later. Ok, I'll stop now :).